### PR TITLE
Add test suite to help new contributors detect regressions

### PR DESCRIPTION
**Pull request by AndreLouisCaron**
I'm hoping to contribute a number of small usability fixes (e.g. see #14) and I would like to do so without fear of breaking your neat package -- I would not want my contributions to be a burden for you.  In order to facilitate this process, I propose to include some basic continuous integration system that help detect regressions.

Build systems are a sensitive topic and different people tend to have strong opinions about them.  If this is too intrusive, feel free to let me know :-)

This change includes a fairly "plain" test suite based on [tox](https://tox.readthedocs.io/en/latest/) and [pytest](https://docs.pytest.org/en/latest/).  It includes [a linter](http://flake8.pycqa.org/en/latest/) and [code coverage](https://coverage.readthedocs.io/en/coverage-4.4.2/) measurement.

The new tests reach 100% code coverage and enforce future builds to do the same (see `coverage report --fail-under=100` in `tox.ini`).

The builds use pip-tools to pin requirements in order to make builds reproducible over time (upstream updates will not cause the build to break).

I tried to make as few changes to the code as I could in order to make it straightforward to review.  I still made a few changes here and there when it simplified the test suite.

If you accept this contribution, I can help you set up a continuous integration build using [Travis-CI](https://travis-ci.org/) (it's free and takes less than 5 minutes to set up).  From then on, you would get a build report automatically any time someone sends a pull request!

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+
+
+[flake8]
+ignore =
+  E111  # indentation is not a multiple of four
```
Comment (AndreLouisCaron) [inline .flake8:6]: **Note to reviewers**: I would like to submit a separate pull request that fixes all linter issues.

File: flamegraph/flamegraph.py [hunk 2]
```diff
@@ -18,21 +19,20 @@ def get_thread_name(ident):
 def default_format_entry(threadname, fname, line, fun, fmt='%(threadname)s`%(fun)s'):
   return fmt % locals()
 
-def create_flamegraph_entry(thread_id, frame, format_entry, collapse_recursion=False):
+def create_flamegraph_entry(thread_id, stack, format_entry, collapse_recursion=False):
```
Comment (AndreLouisCaron) [inline flamegraph/flamegraph.py:22]: **Note to reviewers**: accepting a stack parsed by the caller makes this function easier to unit test.

File: flamegraph/flamegraph.py [hunk 3]
```diff
@@ -109,6 +110,16 @@ def start_profile_thread(fd, interval=0.001, filter=None, format_entry=default_f
   profile_thread.start()
   return profile_thread
 
+@contextlib.contextmanager
+def profile(*args, **kwds):
+  """Profile active threads in the background."""
```
Comment (AndreLouisCaron) [inline flamegraph/flamegraph.py:115]: **Note to reviewers**: this context manager is extacted from code in `main()`.  It makes it easier to unit test the profiling thread without executing an external script.

File: setup.py [hunk 4]
```diff
@@ -18,7 +18,12 @@
       'License :: Public Domain',
       'Programming Language :: Python',
       'Topic :: Software Development :: Debuggers',
-      ]
-    )
+    ],
+    entry_points={
+        'console_scripts': [
+            'flamegraph = flamegraph.__main__:main',
```
Comment (AndreLouisCaron) [inline setup.py:24]: **Note to reviewers**: this allows running from the command-line as `flamegraph` (in addition to the existing `python -m flamegraph`).

**Comment by evanhempel** [comment]
Thanks for contributing! Just starting to take a look at this. I'll probably be rather slow due to other demands on my time, but I appreciate it and am not ignoring you!

File: flamegraph/flamegraph.py [hunk 5]
```diff
@@ -81,7 +83,6 @@ def _write_results(self):
       self._written = True
       for key in sorted(self._stats.keys()):
         self._fd.write('%s %d\n' % (key, self._stats[key]))
-      self._fd.close()
```
Comment (evanhempel) [inline flamegraph/flamegraph.py:84]: Why this deletion?

File: .flake8 (hunk 1 above)
Comment (evanhempel) [reply .flake8:6]: FWIW, I don't really like the 4 space indentation from PEP-8, so most of this is intentional.

Comment (AndreLouisCaron) [reply .flake8:6]: No problem.  I might be able to fix some of the others though.

File: flamegraph/flamegraph.py (hunk 5 above)
Comment (AndreLouisCaron) [reply flamegraph/flamegraph.py:84]: The tests that rely on `capsys` cannot pass otherwise: closing the stream prevents reading from it near the end of test.

I'm actually curious to know why you put this in the first place since standard output is closed automatically :-)

Comment (nickodell) [reply flamegraph/flamegraph.py:84]: @AndreLouisCaron This doesn't look correct. If the user invokes flamegraph like `python -m flamegraph -o profile.log`, then `self._fd` will be a file, which won't be closed automatically.

Comment (AndreLouisCaron) [reply flamegraph/flamegraph.py:84]: Aren't all file descriptors are closed automatically (by the OS) when the process exits?

Comment (nickodell) [reply flamegraph/flamegraph.py:84]: Yes, but this doesn't necessarily mean that all buffers are flushed.

Python buffers writes to files. In the following test program, calling the write() method on the file does not result in a write() syscall until the file is closed.

```python
import time
f = open("foo", "wt")
f.write("test")
time.sleep(10)
f.close()
```

You can test this with

```bash
$ strace -e write,open,close,openat python3 test.py
```

The write() syscall happens after the ten-second wait, despite the write() method happening before the sleep.

In CPython, a close() will happen eventually no matter what. (Even if you comment out `f.close()`, the close syscall will still happen.) However, I think that's an implementation-defined behavior, and not something you should rely on.

See also [Is explictly closing files important?](https://stackoverflow.com/questions/7395542/is-explicitly-closing-files-important), [Python docs](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files)
